File: apps/api/src/services/accountStatus.ts
import crypto from 'node:crypto';
import type { FastifyBaseLogger } from 'fastify';
import { cwGraphqlRequest } from './craftworldClient';

const ACCOUNT_STATUS_TTL = 5_000;

const statusCache = new Map<string, { ts: number; value: any }>();

const ACCOUNT_STATUS_QUERY = `
  query AccountStatus {
    account {
      power
      powerMillisecondsUntilRefill
      powerLastRefill
      updatedAt
    }
  }
`;

const formatHmsFromSeconds = (seconds: number) => {
  const value = Math.max(0, Math.trunc(seconds));
  const h = Math.floor(value / 3600);
  const m = Math.floor((value % 3600) / 60);
  const s = value % 60;
  return `${String(h).padStart(2, '0')}:${String(m).padStart(2, '0')}:${String(s).padStart(2, '0')}`;
};

const tokenCacheKey = (token: string) =>
  crypto.createHash('sha256').update(token, 'utf8').digest('hex');

export async function fetchAccountStatus(jwtToken: string, logger?: FastifyBaseLogger) {
  const now = Date.now();
  const key = tokenCacheKey(jwtToken || '');
  const cached = statusCache.get(key);
  if (cached && now - cached.ts < ACCOUNT_STATUS_TTL) {
    return { ...cached.value };
  }

  const upstream = await cwGraphqlRequest(ACCOUNT_STATUS_QUERY, undefined, jwtToken, logger);
  const body = upstream.body || {};
  const errors = body.errors || [];
  if (!upstream.ok || errors.length) {
    const errorMessage = errors.length ? `Craft World error: ${JSON.stringify(errors)}` : 'Craft World auth failed.';
    const payload = {
      ok: false,
      auth: 'missing_or_invalid',
      power: null,
      msUntilRefill: null,
      refillSeconds: null,
      refillHMS: null,
      primaryWallet: null,
      powerLastRefill: null,
      updatedAt: null,
      error: errorMessage,
      rawErrors: errors,
    };
    statusCache.set(key, { ts: now, value: payload });
    return payload;
  }

  const account = body.data?.account || {};
  const ms = Number(account.powerMillisecondsUntilRefill || 0);
  const refillSeconds = Math.max(0, Math.trunc(ms / 1000));

  const payload = {
    ok: true,
    auth: 'ok',
    power: Number(account.power || 0),
    msUntilRefill: ms,
    refillSeconds,
    refillHMS: formatHmsFromSeconds(refillSeconds),
    primaryWallet: null,
    powerLastRefill: account.powerLastRefill ?? null,
    updatedAt: account.updatedAt ?? null,
  };
  statusCache.set(key, { ts: now, value: payload });
  return payload;
}

File: craftworld_tools/services/account_status.py
"""Account status cache and response shaping helpers."""

from __future__ import annotations

import time
from typing import Any, Dict

from craftworld_api import fetch_account_status

from craftworld_tools.utils.auth import mask_token, normalize_cw_token, token_cache_key
from craftworld_tools.utils.formatting import format_hms_from_seconds

ACCOUNT_STATUS_CACHE: Dict[str, Dict[str, Any]] = {}
ACCOUNT_STATUS_CACHE_TTL = 5.0


def fetch_account_status_for_token(jwt_token: str, logger: Any = None) -> Dict[str, Any]:
    jwt_token = normalize_cw_token(jwt_token) or ""
    now = time.time()
    key = token_cache_key(jwt_token)
    cached_entry = ACCOUNT_STATUS_CACHE.get(key) or {}
    cached_payload = cached_entry.get("value")
    cached_ts = float(cached_entry.get("ts") or 0.0)
    if cached_payload and (now - cached_ts) < ACCOUNT_STATUS_CACHE_TTL:
        return dict(cached_payload)

    if logger is not None:
        logger.debug(
            "Account status token present=%s len=%s token=%s",
            bool(jwt_token),
            len(jwt_token or ""),
            mask_token(jwt_token),
        )

    try:
        account = fetch_account_status(bearer_token=jwt_token)
    except Exception as exc:
        err_text = str(exc)
        if logger is not None and "GraphQL errors:" in err_text:
            logger.warning("Craft World GraphQL error for /api/account_status: %s", err_text)
        response_payload = {
            "ok": False,
            "auth": "missing_or_invalid",
            "power": None,
            "msUntilRefill": None,
            "refillSeconds": None,
            "refillHMS": None,
            "primaryWallet": None,
            "powerLastRefill": None,
            "updatedAt": None,
            "error": f"Craft World auth failed: {err_text}",
            "rawErrors": [],
        }
        ACCOUNT_STATUS_CACHE[key] = {"ts": now, "value": dict(response_payload)}
        return response_payload

    ms = int(account.get("powerMillisecondsUntilRefill") or 0)
    refill_seconds = max(0, ms // 1000)

    response_payload = {
        "ok": True,
        "auth": "ok",
        "power": int(account.get("power") or 0),
        "msUntilRefill": ms,
        "refillSeconds": refill_seconds,
        "refillHMS": format_hms_from_seconds(refill_seconds),
        "primaryWallet": None,
        "powerLastRefill": account.get("powerLastRefill"),
        "updatedAt": account.get("updatedAt"),
    }
    ACCOUNT_STATUS_CACHE[key] = {"ts": now, "value": dict(response_payload)}
    return response_payload
